### PR TITLE
Inherit NET_BIND_SERVICE from IC to Nginx

### LINKED ISSUE #3714: SysctlForbidden on Deployment when controller.hostNetwork = true
**Describe the bug**
With the advent of #3573 it appears that any deployment using hostNetwork=true on a deployment (not sure about daemonset) will now encounter an error with 

```
forbidden sysctl: "net.ipv4.ip_unprivileged_port_start" not allowed with host net enabled
```

**To Reproduce**
1. Deploy using the HELM charts with controller.hostNetwork set to true

**Expected behavior**
nginx pods start up successfully

**Your environment**
* Version of the Ingress Controller - 3.1.0
* Version of Kubernetes: 1.24
* Kubernetes platform (e.g. Mini-kube or GCP): Microk8s 
* Using NGINX or NGINX Plus: nginx

Kube Version Info:
```
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.2", GitCommit:"fc04e732bb3e7198d2fa44efa5457c7c6f8c0f5b", GitTreeState:"clean", BuildDate:"2023-02-22T13:39:03Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"24+", GitVersion:"v1.24.12-2+36e803be0d732b", GitCommit:"36e803be0d732b9da15f6568ed23e5c1929f8701", GitTreeState:"clean", BuildDate:"2023-03-17T18:58:08Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
```

Pinning the HELM chart version to 0.16.2 resolves the problem

### PR DESCRIPTION
**Pull request by sigv**
### Proposed changes

8be0144: _Rework port binding logic without privileges_ caused issues for host networking configurations. The Kubernetes documentation states that the `net.*` sysctls can be used with container networking, which was misinterpreted.

This commit reverts the change, bringing back NET_BIND_SERVICE to the Nginx process, as well as reverts the libcap package removal done in a later commit.

In order to avoid privilege escalation being re-introduced, the IC process is also receiving NET_BIND_SERVICE, so that it can be inherited over to Nginx.

This change aims to restore host networking as functional for the Helm chart. A future change is recommended to harden security for the IC process (to drop the capability after executing Nginx) as well as Nginx itself (to drop the capability after binding).

OBS! To use a 3.1.0 image, you should manually install the `setcap` binary and add `+ep` on `/nginx-ingress` and `+eip` on `nginx` binary.

Resolves #3714.
Effectively reverts #3573 and #3616.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by tstraley** [comment]
IMO this is a better solution anyway because the [Kubernetes Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards) allows for adding the `NET_BIND_SERVICE` capability within the [restricted policy](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted), whereas the `sysctls` `net.ipv4.ip_unprivileged_port_start` is only permitted in the [baseline policy](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline).

I _think_ this means NIC will be fully compatible with the restricted policy after this, if I'm not missing anything.

**Comment by sigv** [comment]
> IMO this is a better solution anyway because the [Kubernetes Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards) allows for adding the `NET_BIND_SERVICE` capability within the [restricted policy](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted), whereas the `sysctls` `net.ipv4.ip_unprivileged_port_start` is only permitted in the [baseline policy](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline).
> 
> I _think_ this means NIC will be fully compatible with the restricted policy after this, if I'm not missing anything.

From what I see, `net.ipv4.ip_unprivileged_port_start` is allowed in Baseline policy as it is categorized as a safe sysctl.
However I am not seeing any mention of it being blocked in Restricted policy.

Keep in mind that Restricted policy inherits Baseline policy.

Am I missing something here?

**Comment by tstraley** [comment]
> From what I see, `net.ipv4.ip_unprivileged_port_start` is allowed in Baseline policy as it is categorized as a safe sysctl. However I am not seeing any mention of it being blocked in Restricted policy.
> 
> Keep in mind that Restricted policy inherits Baseline policy.


Oh, I see - that wasn't clear to me. Thanks.

**Comment by sigv** [comment]
> Oh, I see - that wasn't clear to me. Thanks.

The first line of Controls states _Everything from the baseline profile._
But I do agree the Documentation can be a tricky beast.

As to why this solution is objectively cleaner, it allows dropping the capability in turn locking down ports 80/443.

**Comment by sigv** [comment]
`VirtualServerRoute.k8s.nginx.org` strangeness when running tests locally, not sure how this change could cause that.

```text
suite/test_v_s_route.py::TestVirtualServerRouteValidation::test_openapi_validation_flow[crd_ingress_controller0-v_s_route_setup0]
-------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------
ERROR    root:vs_vsr_resources_utils.py:188 Failed with exception while patching VirtualServerRoute: backend2
Traceback (most recent call last):
  File "/home/azureuser/kubernetes-ingress/tests/suite/utils/vs_vsr_resources_utils.py", line 183, in patch_v_s_route_from_yaml
    custom_objects.patch_namespaced_custom_object(
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/api/custom_objects_api.py", line 2941, in patch_namespaced_custom_object
    return self.patch_namespaced_custom_object_with_http_info(group, version, namespace, plural, name, body, **kwargs)  # noqa: E501
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/api/custom_objects_api.py", line 3072, in patch_namespaced_custom_object_with_http_info
    return self.api_client.call_api(
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/api_client.py", line 407, in request
    return self.rest_client.PATCH(url,
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/rest.py", line 296, in PATCH
    return self.request("PATCH", url,
  File "/home/azureuser/venv/lib/python3.10/site-packages/kubernetes/client/rest.py", line 235, in request
    raise ApiException(http_resp=r)
kubernetes.client.exceptions.ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Audit-Id': '514ffa11-3d33-4614-bba9-425893ba7e42', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': 'f0b5ec4f-ce76-4ca2-abd4-790b16e418db', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'e3033d16-6204-451c-8cb3-8fe069cde889', 'Date': 'Mon, 03 Apr 2023 20:32:48 GMT', 'Content-Length': '584'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"VirtualServerRoute.k8s.nginx.org \"backend2\" is invalid: spec.subroutes[0].action.pass: Invalid value: \"integer\": spec.subroutes[0].action.pass in body must be of type string: \"integer\"","reason":"Invalid","details":{"name":"backend2","group":"k8s.nginx.org","kind":"VirtualServerRoute","causes":[{"reason":"FieldValueTypeInvalid","message":"Invalid value: \"integer\": spec.subroutes[0].action.pass in body must be of type string: \"integer\"","field":"spec.subroutes[0].action.pass"}]},"code":422}
```

**Comment by sigv** [comment]
It seems the edge image is failing `VirtualServerRoute.k8s.nginx.org` tests when I run it, and so is my local testing. Looks good otherwise.

**Comment by sigv** [comment]
I did a rebase, and it looks fine, but Github reports that this PR consists of three commits now. I think that's a Github web UI bug, but let me know if I should try rebasing again.

**Comment by sigv** [comment]
Heads up - the CI failed on

- `debian`: `The repository 'https://pkgs.nginx.com/plus/R28/debian bullseye Release' does not have a Release file.`
- `alpine`: `Ignoring https://pkgs.nginx.com/plus/R28/alpine/v3.17/main: Protocol error`

**Comment by sigv** [comment]
@vepatel, heads up, the Plus repository seems to be unavailable on the CI run!

**Review by ciarams87 — APPROVED** [review]

**Review by jjngx — APPROVED** [review]

**Review by lucacome — CHANGES_REQUESTED** [review]

File: build/Dockerfile [hunk 1]
```diff
@@ -258,6 +264,11 @@ FROM common AS local
 LABEL org.nginx.kic.image.build.version="local"
 
 COPY --link --chown=101:0 nginx-ingress /
+# root is required for `setcap` invocation
+USER 0
+RUN --mount=type=bind,target=/tmp setcap 'cap_net_bind_service=+ep' /nginx-ingress && setcap -v 'cap_net_bind_service=+ep' /nginx-ingress
```
[the remark this answers is not in the document]
Comment (sigv) [reply build/Dockerfile]: Dropped `--mount=type=bind,target=/tmp` from all the `RUN setcap` instructions.

File: build/Dockerfile [hunk 2]
```diff
@@ -237,11 +242,12 @@ ARG IC_VERSION
 ARG TARGETARCH
 
 WORKDIR /go/src/github.com/nginxinc/kubernetes-ingress/
-RUN apk add --no-cache git
+RUN apk add --no-cache git libcap
 RUN --mount=type=bind,target=/go/src/github.com/nginxinc/kubernetes-ingress/ --mount=type=cache,target=/root/.cache/go-build \
 	go mod download
 RUN --mount=type=bind,target=/go/src/github.com/nginxinc/kubernetes-ingress/ --mount=type=cache,target=/root/.cache/go-build \
 	CGO_ENABLED=0 GOOS=linux GOARCH=$TARGETARCH go build -trimpath -ldflags "-s -w -X main.version=${IC_VERSION}" -o /nginx-ingress github.com/nginxinc/kubernetes-ingress/cmd/nginx-ingress
+RUN --mount=type=bind,target=/tmp setcap 'cap_net_bind_service=+ep' /nginx-ingress && setcap -v 'cap_net_bind_service=+ep' /nginx-ingress
```
[the remark this answers is not in the document]
Comment (sigv) [reply build/Dockerfile]: Move the `setcap` call out from `builder` to `container` as requested. ✔️

**Comment by sigv** [comment]
Still hitting issues in CI for accessing the Nginx Plus repository. :/

----

- [Helm Tests (debian-plus, plus)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605275765?pr=3722#step:5:469)
- [Smoke Tests (debian-plus, vs, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300049?pr=3722#step:3:642)
- [Smoke Tests (debian-plus, ts, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300130?pr=3722#step:3:651)
- [Smoke Tests (debian-plus-nap, dos, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300489?pr=3722#step:3:654)
- [Smoke Tests (debian-plus-nap, appprotect, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300590?pr=3722#step:3:668)
```text
E: The repository 'https://pkgs.nginx.com/plus/R28/debian bullseye Release' does not have a Release file.
```

----

- [Smoke Tests (alpine-plus, ingresses, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300214?pr=3722#step:3:439)
- [Smoke Tests (alpine-plus, vsr, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300303?pr=3722#step:3:443)
```text
WARNING: Ignoring https://pkgs.nginx.com/plus/R28/alpine/v3.17/main: Protocol error
ERROR: unable to select packages:
  nginx-plus (no such package):
    required by: world[nginx-plus]
  nginx-plus-module-njs (no such package):
    required by: world[nginx-plus-module-njs]
  nginx-plus-module-opentracing (no such package):
    required by: world[nginx-plus-module-opentracing]
```

----

All of these jobs report warnings for `nginx-repo.tls` secrets at start of test run:
```text
Warning: nginx-repo.crt= is not a valid secret
Warning: nginx-repo.key= is not a valid secret
```

And there is [Smoke Tests (debian-plus-nap, dos, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300489?pr=3722#step:3:652) which also reports issue with reading the data:
```
  Could not load client certificate (/etc/ssl/nginx/nginx-repo.crt, SslCert option) or key (/etc/ssl/nginx/nginx-repo.key, SslKey option): Error while reading file. [IP: 52.59.52.27 443]
```

**Comment by sigv** [comment]
@ciarams87 ping ☝🏻

**Comment by ciarams87** [comment]
> Still hitting issues in CI for accessing the Nginx Plus repository. 😕
> 
> * [Helm Tests (debian-plus, plus)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605275765?pr=3722#step:5:469)
> * [Smoke Tests (debian-plus, vs, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300049?pr=3722#step:3:642)
> * [Smoke Tests (debian-plus, ts, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300130?pr=3722#step:3:651)
> * [Smoke Tests (debian-plus-nap, dos, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300489?pr=3722#step:3:654)
> * [Smoke Tests (debian-plus-nap, appprotect, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300590?pr=3722#step:3:668)
> 
> ```
> E: The repository 'https://pkgs.nginx.com/plus/R28/debian bullseye Release' does not have a Release file.
> ```
> 
>  * [Smoke Tests (alpine-plus, ingresses, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300214?pr=3722#step:3:439)
> * [Smoke Tests (alpine-plus, vsr, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300303?pr=3722#step:3:443)
> 
> ```
> WARNING: Ignoring https://pkgs.nginx.com/plus/R28/alpine/v3.17/main: Protocol error
> ERROR: unable to select packages:
>   nginx-plus (no such package):
>     required by: world[nginx-plus]
>   nginx-plus-module-njs (no such package):
>     required by: world[nginx-plus-module-njs]
>   nginx-plus-module-opentracing (no such package):
>     required by: world[nginx-plus-module-opentracing]
> ```
> 
> All of these jobs report warnings for `nginx-repo.tls` secrets at start of test run:
> 
> ```
> Warning: nginx-repo.crt= is not a valid secret
> Warning: nginx-repo.key= is not a valid secret
> ```
> 
> While most Debian runs have a generic "could not get a Release file" error.. There is [Smoke Tests (debian-plus-nap, dos, 1.27.1)](https://github.com/nginxinc/kubernetes-ingress/actions/runs/4829317490/jobs/8605300489?pr=3722#step:3:652) which also reports issue with reading the secret data:
> 
> ```
>   Could not load client certificate (/etc/ssl/nginx/nginx-repo.crt, SslCert option) or key (/etc/ssl/nginx/nginx-repo.key, SslKey option): Error while reading file.
> ```

@sigv Running the pipeline again now. Sorry for the inconvenience! 

(As an explanation, PRs from external contributors cannot access the build secrets. If there are no changes to the base NGINX Plus images, this normally does not cause a problem as we can use cached layers from other builds, but when changes are made which require rebuilding the whole image, as here, a maintainer needs to open a separate branch which does have secret-read access)

**Review by lucacome — APPROVED** [review]
Thanks @sigv !

**Comment by muhammadwasay** [comment]
@sigv I am confused after reading all the comments. Just want to confirm that can nginxinc/kubernetes-ingress avoid the privilege escalation of NET_BIND_SERVICE or not?

**Pull request merged**